## Understanding the Data 

In [3]:
import pyodbc
from sqlalchemy import create_engine

# Database credentials
server = 'techentjan.database.windows.net'
database = 'QAECECRM_jan'
username = 'azureml_user'
password = 'ElevateLivePro12!'

# Connection string
connection_string = f"DRIVER={{ODBC Driver 17 for SQL Server}};SERVER={server};DATABASE={database};UID={username};PWD={password};Encrypt=yes"

# Connect to Azure SQL Database
try:
    conn = pyodbc.connect(connection_string)
    print("Connection successful!")
except Exception as e:
    print(f"Error: {e}")

Connection successful!


In [4]:
import pandas as pd
import numpy as np

In [5]:
query1 = "SELECT PresenterId, AgentId, AccountName, OrganizationName, PresenterTypeId, MailingAddress1, MailingAddress2, MailingCity, MailingStateId, MailingZip, MailingCountryId, PhysicalAddress1, PhysicalAddress2, PhysicalCity, PhysicalStateId, PhysicalZip, PhysicalCountryId, PhysicalGeoLatitude, PhysicalGeoLongitude, Notes, IsActive, CreatedDate, CreatedById, UpdatedDate, UpdatedById, ArchiveDate, ParentPresenterId FROM [dbo].[Presenter]"
conn2 = pyodbc.connect(connection_string)
df_presenter = pd.read_sql(query1, conn2)
df_presenter.head()

PresenterId  AgentId                  AccountName OrganizationName  \
0            0      281                          ASD             None   
1            8      107  COLLEGE OF WILLIAM AND MARY    BETA THETA PI   
2           32      443               WAKE SUPPLY CO             None   
3           82      435            MARRIOTT/RICHMOND             None   
4           85      253                    [Unknown]             None   

   PresenterTypeId                               MailingAddress1  \
0               99                                          None   
1               49                          700 UKROP WAY UNIT M   
2                8  1700 JEFFREY ST                                
3               99                                          None   
4               99                                          None   

             MailingAddress2   MailingCity  MailingStateId MailingZip  ...  \
0                       None          None             999       None  ...   
1  COLLEGE OF WILLIAM & MARY  WILLIAMSBURG              46       None  ...   
2                       None       RALEIGH              28      27610  ...   
3                       None          None             999       None  ...   
4                       None          None             999       None  ...   

   PhysicalGeoLatitude PhysicalGeoLongitude  \
0                  NaN                  NaN   
1              37.2458             -76.6970   
2              35.7455             -78.5467   
3                  NaN                  NaN   
4                  NaN                  NaN   

                                               Notes IsActive  CreatedDate  \
0  \r\n\r\n\r\n\r\nORTH CAROLINA\r\nORTH CAROLINA...    False   2016-11-10   
1                                               None     True   2012-06-04   
2                                               None    False   2003-12-30   
3                                               None    False   2012-04-11   
4                                               None    False   2009-06-12   

  CreatedById             UpdatedDate  UpdatedById  ArchiveDate  \
0         281 2022-04-18 14:13:32.130        281.0         None   
1         107                     NaT          NaN         None   
2         443 2018-06-02 08:47:55.370     999999.0   2018-06-02   
3         435 2018-06-02 08:47:55.370     999999.0   2018-06-02   
4         253 2018-06-02 08:47:55.370     999999.0   2018-06-02   

  ParentPresenterId  
0                 0  
1                 0  
2                 0  
3                 0  
4                 0  

[5 rows x 27 columns]

In [6]:
query2 = "SELECT GenreTypeId, Name, IsActive FROM [dbo].LuGenreType"
df_genre = pd.read_sql(query2, conn)
df_genre.head()


GenreTypeId                     Name  IsActive
0            1  A Cappella/Vocal Groups      True
1            2                 Acrobats      True
2            3                  African      True
3            4       Animal/Circus Acts      True
4            5             Audio Visual      True

In [7]:
query3 = """
SELECT 
    ArtistId,
    Name,
    MailingCountryId,
    PhysicalAddressCity,
    PhysicalAddressStateId,
    PhysicalGeoLatitude,
    PhysicalGeoLongitude,
    IsActive,
    IsExclusive,
    ExclusiveStartDate,
    NationalNetPriceLowerBound,
    NationalNetPriceUpperBound,
    NationalOneOffPrice,
    CreatedDate,
    UpdatedDate,
    IsNational,
    IsNotParticipating
FROM Artist;
"""

# Execute the query and load data into a Pandas DataFrame
artist_df = pd.read_sql(query3, conn)
artist_df.head()

ArtistId                 Name  MailingCountryId PhysicalAddressCity  \
0         2                    O               1.0            RICHMOND   
1         3        GROOVE DELUXE               1.0           ROCK HILL   
2         4  QUINTON PARKER BAND               1.0             ASHLAND   
3         5  BILL BOLEN BAND (4)               1.0         CHAPEL HILL   
4         6       88 KEYS WILSON               1.0     CHARLOTTESVILLE   

   PhysicalAddressStateId  PhysicalGeoLatitude  PhysicalGeoLongitude  \
0                    46.0              37.5770              -77.5388   
1                    41.0              34.8867              -81.0208   
2                    46.0              37.7596              -77.4822   
3                    28.0              35.9672              -79.0444   
4                    46.0              38.0884              -78.5562   

   IsActive  IsExclusive ExclusiveStartDate  NationalNetPriceLowerBound  \
0      True        False                NaT                         NaN   
1      True        False                NaT                         NaN   
2      True        False                NaT                         NaN   
3      True        False                NaT                         NaN   
4      True        False                NaT                         NaN   

   NationalNetPriceUpperBound  NationalOneOffPrice             CreatedDate  \
0                         NaN                  NaN 2018-06-02 04:05:01.677   
1                         NaN                  NaN 2018-06-02 04:05:01.677   
2                         NaN                  NaN 2018-06-02 04:05:01.677   
3                         NaN                  NaN 2018-06-02 04:05:01.683   
4                         NaN                  NaN 2018-06-02 04:05:01.683   

              UpdatedDate  IsNational  IsNotParticipating  
0                     NaT       False               False  
1                     NaT       False               False  
2                     NaT       False               False  
3 2018-09-19 15:21:38.730       False               False  
4 2019-11-18 15:16:17.807       False               False

In [8]:
query4 = "SELECT ActTypeId, Name, IsActive FROM [dbo].LuActType"
df_act = pd.read_sql(query4, conn)
df_act.head()

ActTypeId           Name  IsActive
0          1          Bands      True
1          2      Classical      True
2          3      Comedians      True
3          4            DJs      True
4          5  International      True

In [9]:
query5 = "SELECT * FROM [dbo].ArtistFeedback"
df_feedback = pd.read_sql(query5, conn)
df_feedback.head()

ArtistFeedbackId  ArtistId  PresenterPortalId  ContractId  IsNegative  \
0                 1      3061                  1      100304       False   
1                 2       918                  2      103526       False   
2                 3       154                  3      104617       False   
3                 4        55                  4      105392       False   
4                 5        32                  5      105486       False   

  ReviewDate                                           Feedback  AgentId  \
0 2001-07-19  THE BREEZE BAND ARE THE BEST. THEY ARE NOT ONL...      NaN   
1 1991-01-21                               EVRYTHING WENT GREAT      NaN   
2 1991-02-15                                      BKD MIRAGE 91      NaN   
3 1991-01-09                                    BKD MAINSTAY 91      NaN   
4 1991-01-09                       BAND PLAYS REGULARLY AT CLUB      NaN   

  Response ResponseDate         CreatedDate  CreatedById UpdatedDate  \
0     None          NaT 2006-06-07 09:03:58            0         NaT   
1     None          NaT 2006-06-07 05:42:43            0         NaT   
2     None          NaT 2006-06-07 02:25:48            0         NaT   
3     None          NaT 2006-06-07 00:39:30            0         NaT   
4     None          NaT 2006-06-07 00:19:13            0         NaT   

   UpdatedById  
0          NaN  
1          NaN  
2          NaN  
3          NaN  
4          NaN

In [10]:
query6 = "SELECT * FROM [dbo].BlueCardArtist"
df_bluecardartist = pd.read_sql(query6, conn)
df_bluecardartist.head()

BlueCardArtistId  BlueCardId  ArtistId   Gross  Net      Notes  IsDeleted  \
0                 1           0       136  3250.0  0.0       None      False   
1                 2           0       296  7750.0  0.0  w/ female      False   
2                 3           0     11183     0.0  0.0       None      False   
3                 4           0     15104     0.0  0.0       None      False   
4                 5           0     15104     0.0  0.0       None      False   

              CreatedDate  CreatedById UpdatedDate  UpdatedById  
0 2018-06-02 04:06:42.950            0         NaT          NaN  
1 2018-06-02 04:06:42.950            0         NaT          NaN  
2 2018-06-02 04:06:42.950            0         NaT          NaN  
3 2018-06-02 04:06:42.950            0         NaT          NaN  
4 2018-06-02 04:06:42.950            0         NaT          NaN

In [11]:
query7 = '''SELECT 
    VenueId,
    AgentId,
    Name,
    VenueSettingId,
    IsSettingCovered,
    MailingAddress1,
    MailingAddress2,
    MailingCity,
    MailingStateId,
    MailingZip,
    MailingCountryId,
    PhysicalAddress1,
    PhysicalAddress2,
    PhysicalCity,
    PhysicalStateId,
    PhysicalZip,
    PhysicalCountryId,
    Notes,
    Capacity,
    IsActive,
    PresenterId,
    CreatedDate,
    CreatedById,
    UpdatedDate,
    UpdatedById,
    ArchiveDate,
    VenueManagerId,
    NotifyForLead,
    NotifyForBC,
    NotifyForContract,
    NotifyForApproval
FROM [dbo].Venue'''
df_venue = pd.read_sql(query7, conn)
df_venue.head()

VenueId  AgentId                               Name  VenueSettingId  \
0        1    141.0                     HILTON ATLANTA               3   
1        2    141.0           MARRIOTT MARQUIS ATLANTA               3   
2        3    430.0  Moreland Landing @ Palmetto Bluff               2   
3        4    141.0        EVERGREEN CONFERENCE CENTER               3   
4        5    141.0              MERCEDES-BENZ STADIUM               1   

   IsSettingCovered                          MailingAddress1  \
0             False                  255 COURTLAND STREET NE   
1             False              265 PEACHTREE CENTER AVENUE   
2              True                            Moreland Road   
3             False              EVERGREEN CONFERENCE CENTER   
4             False  1414 ANDREW YOUNG INTERNATIONAL BLVD NW   

    MailingAddress2     MailingCity  MailingStateId MailingZip  ...  \
0                           ATLANTA              11      30303  ...   
1                           ATLANTA              11      30303  ...   
2                          BLUFFTON              41      29910  ...   
3  1 LAKEVIEW DRIVE  STONE MOUNTAIN              11      30086  ...   
4                           ATLANTA              11      30313  ...   

              CreatedDate CreatedById             UpdatedDate UpdatedById  \
0 2018-06-04 11:42:41.587         141                     NaT         NaN   
1 2018-06-04 11:46:28.407         141                     NaT         NaN   
2 2018-06-04 11:52:10.490         445 2023-01-25 09:12:27.047       182.0   
3 2018-06-04 11:55:11.247         141                     NaT         NaN   
4 2018-06-04 11:56:29.607         141 2020-02-05 13:55:16.493       141.0   

   ArchiveDate VenueManagerId  NotifyForLead NotifyForBC  NotifyForContract  \
0         None            NaN           None        None               None   
1         None            NaN           None        None               None   
2         None            NaN           None        None               None   
3         None            NaN           None        None               None   
4         None            NaN           None        None               None   

   NotifyForApproval  
0               None  
1               None  
2               None  
3               None  
4               None  

[5 rows x 31 columns]

## Data Cleaning

In [11]:
tables=[df_presenter,df_genre,artist_df,df_venue,df_genre,df_bluecardartist,df_feedback]

In [12]:
for df in tables:
    print(df.isnull().sum())
    print("-" * 50)  

PresenterId                 0
AgentId                     0
AccountName                 0
OrganizationName        39190
PresenterTypeId             0
MailingAddress1          3352
MailingAddress2         16547
MailingCity              2062
MailingStateId              0
MailingZip               4520
MailingCountryId            0
PhysicalAddress1         3352
PhysicalAddress2        16547
PhysicalCity             2062
PhysicalStateId             0
PhysicalZip              4520
PhysicalCountryId           0
PhysicalGeoLatitude      3930
PhysicalGeoLongitude     3930
Notes                   54001
IsActive                    0
CreatedDate                 0
CreatedById                 0
UpdatedDate             45937
UpdatedById             45937
ArchiveDate             66975
ParentPresenterId           0
dtype: int64
--------------------------------------------------
GenreTypeId    0
Name           0
IsActive       0
dtype: int64
--------------------------------------------------
ArtistId   

In [13]:
tables = {
    "df_presenter": df_presenter,
    "df_genre": df_genre,
    "artist_df": artist_df,
    "df_venue": df_venue,
    "df_bluecardartist": df_bluecardartist,
    "df_feedback": df_feedback
}

for name, df in tables.items():
    print(f"Missing Values Percentage in {name}:")
    missing_percentage = (df.isnull().sum() / len(df)) 
    print(missing_percentage)
    print("-" * 50)


Missing Values Percentage in df_presenter:
PresenterId             0.000000
AgentId                 0.000000
AccountName             0.000000
OrganizationName        0.435875
PresenterTypeId         0.000000
MailingAddress1         0.037281
MailingAddress2         0.184038
MailingCity             0.022934
MailingStateId          0.000000
MailingZip              0.050272
MailingCountryId        0.000000
PhysicalAddress1        0.037281
PhysicalAddress2        0.184038
PhysicalCity            0.022934
PhysicalStateId         0.000000
PhysicalZip             0.050272
PhysicalCountryId       0.000000
PhysicalGeoLatitude     0.043710
PhysicalGeoLongitude    0.043710
Notes                   0.600605
IsActive                0.000000
CreatedDate             0.000000
CreatedById             0.000000
UpdatedDate             0.510916
UpdatedById             0.510916
ArchiveDate             0.744903
ParentPresenterId       0.000000
dtype: float64
--------------------------------------------------


In [14]:

threshold = 0.5


tables = {
    "df_presenter": df_presenter,
    "df_genre": df_genre,
    "artist_df": artist_df,
    "df_venue": df_venue,
    "df_bluecardartist": df_bluecardartist,
    "df_feedback": df_feedback
}


for name, df in tables.items():
    missing_ratio = df.isnull().sum() / len(df)  # Calculate missing percentage
    columns_to_drop = missing_ratio[missing_ratio > threshold].index  # Identify columns > 50% missing
    df.drop(columns=columns_to_drop, inplace=True)  # Drop columns
    print(f"Dropped columns from {name}: {list(columns_to_drop)}")
    print("-" * 50)


Dropped columns from df_presenter: ['Notes', 'UpdatedDate', 'UpdatedById', 'ArchiveDate']
--------------------------------------------------
Dropped columns from df_genre: []
--------------------------------------------------
Dropped columns from artist_df: ['ExclusiveStartDate', 'NationalNetPriceLowerBound', 'NationalNetPriceUpperBound', 'NationalOneOffPrice']
--------------------------------------------------
Dropped columns from df_venue: ['Capacity', 'UpdatedDate', 'UpdatedById', 'ArchiveDate', 'VenueManagerId', 'NotifyForLead', 'NotifyForBC', 'NotifyForContract', 'NotifyForApproval']
--------------------------------------------------
Dropped columns from df_bluecardartist: ['UpdatedDate', 'UpdatedById']
--------------------------------------------------
Dropped columns from df_feedback: ['AgentId', 'Response', 'ResponseDate', 'UpdatedDate', 'UpdatedById']
--------------------------------------------------


In [15]:
tables = {
    "df_presenter": df_presenter,
    "df_genre": df_genre,
    "artist_df": artist_df,
    "df_venue": df_venue,
    "df_bluecardartist": df_bluecardartist,
    "df_feedback": df_feedback
}

for name, df in tables.items():
    print(f"Missing Values Percentage in {name}:")
    missing_percentage = (df.isnull().sum() / len(df)) 
    print(missing_percentage)
    print("-" * 50)

Missing Values Percentage in df_presenter:
PresenterId             0.000000
AgentId                 0.000000
AccountName             0.000000
OrganizationName        0.435875
PresenterTypeId         0.000000
MailingAddress1         0.037281
MailingAddress2         0.184038
MailingCity             0.022934
MailingStateId          0.000000
MailingZip              0.050272
MailingCountryId        0.000000
PhysicalAddress1        0.037281
PhysicalAddress2        0.184038
PhysicalCity            0.022934
PhysicalStateId         0.000000
PhysicalZip             0.050272
PhysicalCountryId       0.000000
PhysicalGeoLatitude     0.043710
PhysicalGeoLongitude    0.043710
IsActive                0.000000
CreatedDate             0.000000
CreatedById             0.000000
ParentPresenterId       0.000000
dtype: float64
--------------------------------------------------
Missing Values Percentage in df_genre:
GenreTypeId    0.0
Name           0.0
IsActive       0.0
dtype: float64
---------------------

In [16]:
# Function to impute missing values
def impute_missing_values(df):
    for col in df.columns:
        if df[col].dtype == "object":  
            df[col].fillna("Unknown", inplace=True)
        elif df[col].dtype in ["int64", "float64"]:  
            if df[col].nunique() < 10:  
                df[col].fillna(df[col].mode()[0], inplace=True)  
            else:
                df[col].fillna(df[col].median(), inplace=True)  
        elif np.issubdtype(df[col].dtype, np.datetime64):  
            df[col].fillna(df[col].min(), inplace=True)  
        elif df[col].dtype == "bool":  
            df[col].fillna(False, inplace=True)  
    return df

In [17]:
# Apply imputation to all DataFrames
for name, df in tables.items():
    tables[name] = impute_missing_values(df)
    print(f"Imputed missing values in {name}")

# Verify no missing values
for name, df in tables.items():
    print(f"Remaining missing values in {name}:")
    print(df.isnull().sum())
    print("-" * 50)

Imputed missing values in df_presenter
Imputed missing values in df_genre
Imputed missing values in artist_df
Imputed missing values in df_venue
Imputed missing values in df_bluecardartist
Imputed missing values in df_feedback
Remaining missing values in df_presenter:
PresenterId             0
AgentId                 0
AccountName             0
OrganizationName        0
PresenterTypeId         0
MailingAddress1         0
MailingAddress2         0
MailingCity             0
MailingStateId          0
MailingZip              0
MailingCountryId        0
PhysicalAddress1        0
PhysicalAddress2        0
PhysicalCity            0
PhysicalStateId         0
PhysicalZip             0
PhysicalCountryId       0
PhysicalGeoLatitude     0
PhysicalGeoLongitude    0
IsActive                0
CreatedDate             0
CreatedById             0
ParentPresenterId       0
dtype: int64
--------------------------------------------------
Remaining missing values in df_genre:
GenreTypeId    0
Name          

In [18]:
for name, df in tables.items():
    print(f"The head of {name} is:\n{df.head()}\n")


The head of df_presenter is:
   PresenterId  AgentId                  AccountName OrganizationName  \
0            0      281                          ASD          Unknown   
1            8      107  COLLEGE OF WILLIAM AND MARY    BETA THETA PI   
2           32      443               WAKE SUPPLY CO          Unknown   
3           82      435            MARRIOTT/RICHMOND          Unknown   
4           85      253                    [Unknown]          Unknown   

   PresenterTypeId                               MailingAddress1  \
0               99                                       Unknown   
1               49                          700 UKROP WAY UNIT M   
2                8  1700 JEFFREY ST                                
3               99                                       Unknown   
4               99                                       Unknown   

             MailingAddress2   MailingCity  MailingStateId MailingZip  ...  \
0                    Unknown       Unknown   

# Collaborative filtering

In [19]:
tables_svd = {
    "df_presenter": df_presenter,
    "artist_df": artist_df,
    "df_feedback": df_feedback,
    "df_genre": df_genre,
    "df_act": df_act
}

# Show the head of each dataframe
for name, df in tables_svd.items():
    print(f"Head of {name}:")
    display(df.head())


Head of df_presenter:


PresenterId  AgentId                  AccountName OrganizationName  \
0            0      281                          ASD          Unknown   
1            8      107  COLLEGE OF WILLIAM AND MARY    BETA THETA PI   
2           32      443               WAKE SUPPLY CO          Unknown   
3           82      435            MARRIOTT/RICHMOND          Unknown   
4           85      253                    [Unknown]          Unknown   

   PresenterTypeId                               MailingAddress1  \
0               99                                       Unknown   
1               49                          700 UKROP WAY UNIT M   
2                8  1700 JEFFREY ST                                
3               99                                       Unknown   
4               99                                       Unknown   

             MailingAddress2   MailingCity  MailingStateId MailingZip  ...  \
0                    Unknown       Unknown             999    Unknown  ...   
1  COLLEGE OF WILLIAM & MARY  WILLIAMSBURG              46    Unknown  ...   
2                    Unknown       RALEIGH              28      27610  ...   
3                    Unknown       Unknown             999    Unknown  ...   
4                    Unknown       Unknown             999    Unknown  ...   

   PhysicalCity PhysicalStateId PhysicalZip PhysicalCountryId  \
0       Unknown             999     Unknown                 1   
1  WILLIAMSBURG              46     Unknown                 1   
2       RALEIGH              28       27610                 1   
3       Unknown             999     Unknown                 1   
4       Unknown             999     Unknown                 1   

   PhysicalGeoLatitude PhysicalGeoLongitude  IsActive  CreatedDate  \
0              35.7373           -80.770107     False   2016-11-10   
1              37.2458           -76.697000      True   2012-06-04   
2              35.7455           -78.546700     False   2003-12-30   
3              35.7373           -80.770107     False   2012-04-11   
4              35.7373           -80.770107     False   2009-06-12   

   CreatedById  ParentPresenterId  
0          281                  0  
1          107                  0  
2          443                  0  
3          435                  0  
4          253                  0  

[5 rows x 23 columns]

Head of artist_df:


ArtistId                 Name  MailingCountryId PhysicalAddressCity  \
0         2                    O               1.0            RICHMOND   
1         3        GROOVE DELUXE               1.0           ROCK HILL   
2         4  QUINTON PARKER BAND               1.0             ASHLAND   
3         5  BILL BOLEN BAND (4)               1.0         CHAPEL HILL   
4         6       88 KEYS WILSON               1.0     CHARLOTTESVILLE   

   PhysicalAddressStateId  PhysicalGeoLatitude  PhysicalGeoLongitude  \
0                    46.0              37.5770              -77.5388   
1                    41.0              34.8867              -81.0208   
2                    46.0              37.7596              -77.4822   
3                    28.0              35.9672              -79.0444   
4                    46.0              38.0884              -78.5562   

   IsActive  IsExclusive             CreatedDate             UpdatedDate  \
0      True        False 2018-06-02 04:05:01.677 2018-06-02 17:21:47.147   
1      True        False 2018-06-02 04:05:01.677 2018-06-02 17:21:47.147   
2      True        False 2018-06-02 04:05:01.677 2018-06-02 17:21:47.147   
3      True        False 2018-06-02 04:05:01.683 2018-09-19 15:21:38.730   
4      True        False 2018-06-02 04:05:01.683 2019-11-18 15:16:17.807   

   IsNational  IsNotParticipating  
0       False               False  
1       False               False  
2       False               False  
3       False               False  
4       False               False

Head of df_feedback:


ArtistFeedbackId  ArtistId  PresenterPortalId  ContractId  IsNegative  \
0                 1      3061                  1      100304       False   
1                 2       918                  2      103526       False   
2                 3       154                  3      104617       False   
3                 4        55                  4      105392       False   
4                 5        32                  5      105486       False   

  ReviewDate                                           Feedback  \
0 2001-07-19  THE BREEZE BAND ARE THE BEST. THEY ARE NOT ONL...   
1 1991-01-21                               EVRYTHING WENT GREAT   
2 1991-02-15                                      BKD MIRAGE 91   
3 1991-01-09                                    BKD MAINSTAY 91   
4 1991-01-09                       BAND PLAYS REGULARLY AT CLUB   

          CreatedDate  CreatedById  
0 2006-06-07 09:03:58            0  
1 2006-06-07 05:42:43            0  
2 2006-06-07 02:25:48            0  
3 2006-06-07 00:39:30            0  
4 2006-06-07 00:19:13            0

Head of df_genre:


GenreTypeId                     Name  IsActive
0            1  A Cappella/Vocal Groups      True
1            2                 Acrobats      True
2            3                  African      True
3            4       Animal/Circus Acts      True
4            5             Audio Visual      True

Head of df_act:


ActTypeId           Name  IsActive
0          1          Bands      True
1          2      Classical      True
2          3      Comedians      True
3          4            DJs      True
4          5  International      True

In [20]:
# Check the column names and data types
print(df_feedback.info())

# Check unique values in key columns to understand data distribution
print("Unique values in IsNegative column:", df_feedback["IsNegative"].unique())

# Check the number of unique presenters and artists in feedback
print("Unique Presenters in Feedback:", df_feedback["PresenterPortalId"].nunique())
print("Unique Artists in Feedback:", df_feedback["ArtistId"].nunique())

# Check if we have meaningful text feedback
print(df_feedback["Feedback"].head(10))

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 47430 entries, 0 to 47429
Data columns (total 9 columns):
 #   Column             Non-Null Count  Dtype         
---  ------             --------------  -----         
 0   ArtistFeedbackId   47430 non-null  int64         
 1   ArtistId           47430 non-null  int64         
 2   PresenterPortalId  47430 non-null  int64         
 3   ContractId         47430 non-null  int64         
 4   IsNegative         47430 non-null  bool          
 5   ReviewDate         47430 non-null  datetime64[ns]
 6   Feedback           47430 non-null  object        
 7   CreatedDate        47430 non-null  datetime64[ns]
 8   CreatedById        47430 non-null  int64         
dtypes: bool(1), datetime64[ns](2), int64(5), object(1)
memory usage: 2.9+ MB
None
Unique values in IsNegative column: [False  True]
Unique Presenters in Feedback: 47216
Unique Artists in Feedback: 4267
0    THE BREEZE BAND ARE THE BEST. THEY ARE NOT ONL...
1                            

In [21]:
# Create a Numerical Ratings Column
# Convert IsNegative into a numerical rating
df_feedback["Rating"] = df_feedback["IsNegative"].apply(lambda x: 0 if x else 1)

# Display updated feedback data
print(df_feedback[["PresenterPortalId", "ArtistId", "Rating"]].head(10))

   PresenterPortalId  ArtistId  Rating
0                  1      3061       1
1                  2       918       1
2                  3       154       1
3                  4        55       1
4                  5        32       1
5                  6       297       1
6                  7       203       1
7                  8        84       1
8                  9      2254       1
9                 10      2181       1


## Constructing the User-Item Interaction Matrix

In [22]:
user_item_matrix = df_feedback.pivot_table(index='PresenterPortalId', 
                                           columns='ArtistId', 
                                           values='Rating', 
                                           fill_value=0)  # Fill missing values with 0 (No interaction)

# Display the matrix
import pandas as pd
import matplotlib.pyplot as plt

# Show dataframe
user_item_matrix.head()


ArtistId           2       3       5       6       7       8       16      \
PresenterPortalId                                                           
0                       0       0       0       0       0       0       0   
1                       0       0       0       0       0       0       0   
2                       0       0       0       0       0       0       0   
3                       0       0       0       0       0       0       0   
4                       0       0       0       0       0       0       0   

ArtistId           17      21      23      ...  102354  102531  102534  \
PresenterPortalId                          ...                           
0                       0       0       0  ...       1       0       0   
1                       0       0       0  ...       0       0       0   
2                       0       0       0  ...       0       0       0   
3                       0       0       0  ...       0       0       0   
4                       0       0       0  ...       0       0       0   

ArtistId           102691  103147  103155  103202  103213  103352  103546  
PresenterPortalId                                                          
0                       0       0       0       0       1       0       0  
1                       0       0       0       0       0       0       0  
2                       0       0       0       0       0       0       0  
3                       0       0       0       0       0       0       0  
4                       0       0       0       0       0       0       0  

[5 rows x 4267 columns]

## SVD (Singular Value Decomposition) for Collaborative Filtering!

In [23]:
from sklearn.decomposition import TruncatedSVD

# Define the number of latent features
n_components = 80  # You can tune this parameter

# Apply Truncated SVD
svd = TruncatedSVD(n_components=n_components, random_state=42)
user_factors = svd.fit_transform(user_item_matrix)
artist_factors = svd.components_

# Display results
print(f"Shape of User Factors (U matrix): {user_factors.shape}")
print(f"Shape of Artist Factors (V matrix): {artist_factors.shape}")
print(f"Explained Variance Ratio: {np.sum(svd.explained_variance_ratio_):.4f}")

Shape of User Factors (U matrix): (47216, 80)
Shape of Artist Factors (V matrix): (80, 4267)
Explained Variance Ratio: 0.4500


The Explained Variance Ratio = 45%, meaning SVD captures about 45% of the variation in user preferences. We might need to increase n_components later for better results

## Compute Predicted Ratings

In [24]:
# Reconstruct the matrix
predicted_ratings = np.dot(user_factors, artist_factors)

# Convert to DataFrame for readability
predicted_ratings_df = pd.DataFrame(predicted_ratings, 
                                    index=user_item_matrix.index, 
                                    columns=user_item_matrix.columns)

# Display first few rows
predicted_ratings_df.head()

ArtistId                 2             3             5             6       \
PresenterPortalId                                                           
0                 -2.387177e-09  7.708342e-05 -4.627905e-09 -1.134064e-09   
1                 -1.515541e-20 -1.629205e-16  2.220070e-21  3.184640e-21   
2                  4.275571e-11 -1.224628e-06 -2.317748e-10  5.177337e-12   
3                 -5.553936e-12  1.794785e-08  9.033720e-12 -2.092563e-13   
4                 -8.381606e-13 -2.204878e-09  8.204516e-13  8.944628e-14   

ArtistId                 7             8             16            17      \
PresenterPortalId                                                           
0                 -3.003064e-08 -6.648920e-10 -1.473540e-13 -1.078488e-02   
1                  4.615757e-21  1.487343e-20 -3.851917e-26 -1.555525e-14   
2                 -2.263593e-10 -4.826327e-11  6.481200e-16  3.648890e-06   
3                 -3.099087e-11 -1.523227e-12  8.304679e-17 -2.777000e-06   
4                 -1.627413e-12  2.759124e-13 -2.364733e-17 -3.381680e-07   

ArtistId                 21            23      ...        102354  \
PresenterPortalId                              ...                 
0                 -1.573149e-09  1.589497e-07  ...  5.317093e-02   
1                  2.857783e-21 -3.183959e-17  ... -3.868752e-15   
2                 -2.771406e-11  3.483004e-08  ... -2.562504e-05   
3                 -4.303667e-13  4.359223e-09  ...  1.198464e-07   
4                  2.429525e-14 -4.994465e-11  ... -1.769129e-07   

ArtistId                 102531        102534        102691        103147  \
PresenterPortalId                                                           
0                 -2.500958e-10 -3.240663e-12  2.142454e-11 -6.788776e-10   
1                 -6.802033e-22 -3.967284e-23 -2.832694e-22  1.104935e-21   
2                  4.748849e-12  1.643866e-13 -3.980932e-13  7.401675e-12   
3                  1.448915e-13  4.815079e-15  1.733957e-13  2.623515e-14   
4                 -4.477055e-14 -1.357506e-15 -7.655815e-15  5.362027e-15   

ArtistId                 103155        103202        103213        103352  \
PresenterPortalId                                                           
0                 -4.535471e-10  3.301357e-13  5.317093e-02  6.927187e-14   
1                  6.854848e-22 -4.469924e-25 -3.868752e-15 -2.281613e-25   
2                  1.219241e-12 -4.284984e-15 -2.562504e-05 -3.627420e-15   
3                  7.264804e-14  1.010147e-16  1.198464e-07  3.434050e-17   
4                  7.368419e-15  1.092469e-18 -1.769129e-07  9.460799e-18   

ArtistId                 103546  
PresenterPortalId                
0                 -3.152391e-13  
1                  4.413733e-26  
2                 -1.629740e-15  
3                 -9.496436e-17  
4                 -1.421081e-17  

[5 rows x 4267 columns]

## Generating Top-N Recommendations for a Given Presenter

In [25]:
def get_top_n_recommendations(presenter_id, n=5):
    """
    Get top N recommended artists for a given presenter based on predicted ratings.
    
    :param presenter_id: ID of the presenter for whom recommendations are generated.
    :param n: Number of recommendations to return.
    :return: DataFrame of top recommended artists with predicted scores.
    """
    if presenter_id not in predicted_ratings_df.index:
        print("Presenter ID not found in data.")
        return None

    # Get predicted ratings for this presenter
    presenter_ratings = predicted_ratings_df.loc[presenter_id]

    # Sort artists by predicted rating in descending order
    top_artists = presenter_ratings.sort_values(ascending=False).head(n)

    # Convert to DataFrame
    recommendations = pd.DataFrame({"ArtistId": top_artists.index, "PredictedRating": top_artists.values})

    return recommendations

In [26]:
# Get top 5 recommendations for PresenterPortalId = 10
presenter_id = 10
top_recommendations = get_top_n_recommendations(presenter_id, n=5)

# Display results
top_recommendations

ArtistId  PredictedRating
0        77         0.000025
1      3003         0.000023
2      1104         0.000014
3      8701         0.000014
4        37         0.000011

The predicted ratings are quite small, likely due to sparse feedback data.
We can further filter or enhance recommendations using genre, location, or feedback scores.

## Improving the performance
### Normalize Predicted Ratings

In [27]:
from sklearn.preprocessing import MinMaxScaler

# Normalize predicted ratings to a 0-1 scale
scaler = MinMaxScaler()
normalized_ratings = scaler.fit_transform(predicted_ratings)

# Convert back to a DataFrame
normalized_ratings_df = pd.DataFrame(normalized_ratings, 
                                     index=predicted_ratings_df.index, 
                                     columns=predicted_ratings_df.columns)

# Display the first few rows
print(normalized_ratings_df.head())

ArtistId             2         3         5         6         7         8       \
PresenterPortalId                                                               
0                  0.464120  1.000000  0.343387  0.191725  0.194859  0.370378   
1                  0.527295  0.516971  0.446763  0.461578  0.505693  0.413311   
2                  0.528426  0.509297  0.441585  0.462810  0.503350  0.410194   
3                  0.527148  0.517084  0.446964  0.461529  0.505373  0.413212   
4                  0.527272  0.516957  0.446781  0.461600  0.505676  0.413328   

ArtistId             16        17        21        23      ...    102354  \
PresenterPortalId                                          ...             
0                  0.263163  0.287736  0.030252  0.000017  ...  1.000000   
1                  0.468206  0.551469  0.342953  0.000017  ...  0.070793   
2                  0.469108  0.551558  0.337444  0.000017  ...  0.070345   
3                  0.468322  0.551401  0.342868  0.0

In [28]:
# Retrieve Adjusted Top-N Recommendations
def get_adjusted_top_n_recommendations(presenter_id, n=5):
    """
    Get top N recommended artists for a given presenter using normalized ratings.
    
    :param presenter_id: ID of the presenter for whom recommendations are generated.
    :param n: Number of recommendations to return.
    :return: DataFrame of top recommended artists with predicted scores.
    """
    if presenter_id not in normalized_ratings_df.index:
        print("Presenter ID not found in data.")
        return None

    # Get normalized predicted ratings for this presenter
    presenter_ratings = normalized_ratings_df.loc[presenter_id]

    # Sort artists by predicted rating in descending order
    top_artists = presenter_ratings.sort_values(ascending=False).head(n)

    # Convert to DataFrame
    recommendations = pd.DataFrame({"ArtistId": top_artists.index, "NormalizedRating": top_artists.values})

    return recommendations

In [29]:
# Get top 5 recommendations for PresenterPortalId = 10
presenter_id = 10
adjusted_top_recommendations = get_adjusted_top_n_recommendations(presenter_id, n=5)

# Display results
print(adjusted_top_recommendations)

   ArtistId  NormalizedRating
0      5349          0.835598
1     14809          0.806696
2      2378          0.804622
3      6984          0.799171
4      1328          0.787272


In [30]:
# Merge Recommendations with Artist Names
adjusted_top_recommendations = adjusted_top_recommendations.merge(artist_df[["ArtistId", "Name"]], on="ArtistId", how="left")

# Rename column for clarity
adjusted_top_recommendations.rename(columns={"Name": "ArtistName"}, inplace=True)

# Display updated recommendations
adjusted_top_recommendations


ArtistId  NormalizedRating              ArtistName
0      5349          0.835598             GATOR GUMBO
1     14809          0.806696    BOOK IT INCORPORATED
2      2378          0.804622  PIANIST - PAUL BARTSCH
3      6984          0.799171                JAY MOHR
4      1328          0.787272                   MISTA

##  Improving Model Performance (Hyperparameter Tuning & Enhancements)
### Tune n_components for SVD

In [31]:
from sklearn.decomposition import TruncatedSVD
import numpy as np

# Test different n_components
best_explained_variance = 0
best_n = 0
explained_variances = {}

for n in [100, 120, 150, 180, 200]:  # Different values to test
    svd = TruncatedSVD(n_components=n, random_state=42)
    user_factors = svd.fit_transform(user_item_matrix)
    artist_factors = svd.components_
    
    explained_variance = np.sum(svd.explained_variance_ratio_)
    explained_variances[n] = explained_variance
    
    print(f"n_components={n}, Explained Variance={explained_variance:.2%}")
    
    if explained_variance > best_explained_variance:
        best_explained_variance = explained_variance
        best_n = n

print(f"\nBest n_components={best_n} with Explained Variance={best_explained_variance:.2%}")

n_components=100, Explained Variance=48.61%
n_components=120, Explained Variance=51.64%
n_components=150, Explained Variance=55.40%
n_components=180, Explained Variance=58.48%
n_components=200, Explained Variance=60.25%

Best n_components=200 with Explained Variance=60.25%


In [33]:
# Compute sparsity
num_interactions = np.count_nonzero(user_item_matrix)
total_possible_interactions = user_item_matrix.shape[0] * user_item_matrix.shape[1]
sparsity = 1 - (num_interactions / total_possible_interactions)

print(f" Data Sparsity: {sparsity:.4%} of the matrix is empty (no interaction).")


 Data Sparsity: 99.9774% of the matrix is empty (no interaction).


In [34]:
# Compute the number of interactions per presenter and artist
presenter_interactions = user_item_matrix.astype(bool).sum(axis=1)
artist_interactions = user_item_matrix.astype(bool).sum(axis=0)

# Set threshold for minimum interactions
min_interactions = 3  

# Filter presenters and artists with sufficient interactions
active_presenters = presenter_interactions[presenter_interactions >= min_interactions].index
active_artists = artist_interactions[artist_interactions >= min_interactions].index

# Filter the user-item matrix
filtered_user_item_matrix = user_item_matrix.loc[active_presenters, active_artists]

print(f" New Matrix Shape: {filtered_user_item_matrix.shape} (After filtering)")

 New Matrix Shape: (2, 1751) (After filtering)


In [35]:
# Convert explicit ratings into implicit interactions
implicit_feedback_matrix = user_item_matrix.copy()

# Assign weights based on interaction frequency
implicit_feedback_matrix = np.log1p(implicit_feedback_matrix)  # Log scaling to prevent bias from large counts

# Normalize between 0 and 1
from sklearn.preprocessing import MinMaxScaler

scaler = MinMaxScaler()
implicit_feedback_matrix = pd.DataFrame(
    scaler.fit_transform(implicit_feedback_matrix),
    index=user_item_matrix.index,
    columns=user_item_matrix.columns
)

print(f" Implicit Feedback Matrix Shape: {implicit_feedback_matrix.shape}")

 Implicit Feedback Matrix Shape: (47216, 4267)


In [36]:
from sklearn.decomposition import TruncatedSVD

# Set optimal n_components based on previous tuning
optimal_n_components = 200  # We can tune this later if needed

# Apply Truncated SVD
svd = TruncatedSVD(n_components=optimal_n_components, random_state=42)
user_factors = svd.fit_transform(implicit_feedback_matrix)
artist_factors = svd.components_

# Compute explained variance
explained_variance = np.sum(svd.explained_variance_ratio_) * 100
print(f" New Explained Variance: {explained_variance:.2f}%")

 New Explained Variance: 60.25%


In [39]:
explained_variances = {}
for n in [250, 300, 350, 400]:  # Increasing components
    svd = TruncatedSVD(n_components=n, random_state=42)
    svd.fit(implicit_feedback_matrix)

    explained_variance = np.sum(svd.explained_variance_ratio_) * 100
    explained_variances[n] = explained_variance
    print(f" Testing n_components={n} | Explained Variance={explained_variance:.2f}%")

# Find the best value
best_n = max(explained_variances, key=explained_variances.get)
best_explained_variance = explained_variances[best_n]
print(f"\n Best n_components={best_n} with Explained Variance={best_explained_variance:.2f}%")

 Testing n_components=250 | Explained Variance=64.00%
 Testing n_components=300 | Explained Variance=67.17%
 Testing n_components=350 | Explained Variance=69.82%
 Testing n_components=400 | Explained Variance=72.12%

 Best n_components=400 with Explained Variance=72.12%


In [40]:
# Apply SVD with best n_components
svd = TruncatedSVD(n_components=400, random_state=42)
user_factors = svd.fit_transform(implicit_feedback_matrix)
artist_factors = svd.components_

# Reconstruct the matrix
predicted_ratings = np.dot(user_factors, artist_factors)

# Convert back to DataFrame
predicted_ratings_df = pd.DataFrame(predicted_ratings, 
                                    index=implicit_feedback_matrix.index, 
                                    columns=implicit_feedback_matrix.columns)

print(" Model retrained with n_components=400. Predictions updated!")


 Model retrained with n_components=400. Predictions updated!


In [41]:
def get_top_n_recommendations(presenter_id, n=5):
    """
    Get top N recommended artists for a given presenter based on predicted ratings.
    
    :param presenter_id: ID of the presenter for whom recommendations are generated.
    :param n: Number of recommendations to return.
    :return: DataFrame of top recommended artists with predicted scores.
    """
    if presenter_id not in predicted_ratings_df.index:
        print("Presenter ID not found in data.")
        return pd.DataFrame()

    # Get top-N artists
    top_artists = predicted_ratings_df.loc[presenter_id].nlargest(n).reset_index()
    top_artists.columns = ["ArtistId", "PredictedRating"]
    
    return top_artists

# Get top 5 recommendations for PresenterPortalId=10
presenter_id = 10
top_recommendations = get_top_n_recommendations(presenter_id, n=5)

print(" Top 5 Recommended Artists:")
print(top_recommendations)

 Top 5 Recommended Artists:
   ArtistId  PredictedRating
0      2181         0.193504
1     30263         0.095348
2      2049         0.077884
3       113         0.074200
4     11234         0.067995


In [43]:
from sklearn.metrics import mean_squared_error, mean_absolute_error
import numpy as np

# Ensure matrices have the same shape
actual_ratings = implicit_feedback_matrix.to_numpy()
predicted_ratings = predicted_ratings_df.to_numpy()

# Mask non-zero values (actual interactions only)
mask = actual_ratings > 0  

# Compute RMSE & MAE
rmse = np.sqrt(mean_squared_error(actual_ratings[mask], predicted_ratings[mask]))
mae = mean_absolute_error(actual_ratings[mask], predicted_ratings[mask])

print(f"RMSE: {rmse:.4f}")
print(f"MAE: {mae:.4f}")


 RMSE: 0.5146
 MAE: 0.2777


In [45]:
def precision_at_k(predictions_df, actual_interactions, k=5):
    """
    Compute Precision@K for top-k recommendations.
    
    :param predictions_df: DataFrame of predicted ratings.
    :param actual_interactions: Actual user-item interactions.
    :param k: Number of top items to consider.
    :return: Precision@K score.
    """
    total_precision = 0
    num_users = predictions_df.shape[0]

    for user_idx in range(num_users):
        # Get top-K predictions
        top_k_predicted = predictions_df.iloc[user_idx].nlargest(k).index.tolist()

        # Get actual positive interactions (convert Series to array)
        actual_liked = set(np.where(actual_interactions.iloc[user_idx].values > 0)[0])

        # Compute precision
        if actual_liked:
            precision = len(set(top_k_predicted) & actual_liked) / k
            total_precision += precision
    
    return total_precision / num_users if num_users > 0 else 0

# Compute Precision@5
precision_5 = precision_at_k(predicted_ratings_df, implicit_feedback_matrix, k=5)
print(f"Precision@5: {precision_5:.4f}")

Precision@5: 0.0001


Precision@5 is extremely low (0.0001), meaning that the model’s top recommendations do not match actual user preferences well.
However, RMSE (0.5146) and MAE (0.2777) are reasonable, indicating the predicted ratings align well with the dataset's structure.

Since the dataset has 99.97% sparsity, this is expected. We need better ranking for recommendations rather than just predicting ratings.

# LightFM

In [10]:
print(df_presenter.isnull().sum())
print(artist_df.isnull().sum())
print(df_feedback.isnull().sum())

PresenterId                 0
AgentId                     0
AccountName                 0
OrganizationName        39190
PresenterTypeId             0
MailingAddress1          3352
MailingAddress2         16547
MailingCity              2062
MailingStateId              0
MailingZip               4520
MailingCountryId            0
PhysicalAddress1         3352
PhysicalAddress2        16547
PhysicalCity             2062
PhysicalStateId             0
PhysicalZip              4520
PhysicalCountryId           0
PhysicalGeoLatitude      3930
PhysicalGeoLongitude     3930
Notes                   54001
IsActive                    0
CreatedDate                 0
CreatedById                 0
UpdatedDate             45937
UpdatedById             45937
ArchiveDate             66975
ParentPresenterId           0
dtype: int64
ArtistId                          0
Name                              0
MailingCountryId               2655
PhysicalAddressCity            3577
PhysicalAddressStateId         35

In [12]:
print("Unique Presenters:", df_presenter["PresenterId"].nunique())
print("Unique Artists:", artist_df["ArtistId"].nunique())
print("Unique Feedback Records:", df_feedback["ArtistId"].nunique())


Unique Presenters: 89911
Unique Artists: 23590
Unique Feedback Records: 4267


In [13]:
# Dropping irrelevant columns with too many missing values
df_presenter_cleaned = df_presenter.drop(columns=["Notes", "UpdatedDate", "UpdatedById", "ArchiveDate"])
artist_df_cleaned = artist_df.drop(columns=["ExclusiveStartDate", "UpdatedDate"])
df_feedback_cleaned = df_feedback.drop(columns=["Response", "ResponseDate", "UpdatedDate", "UpdatedById", "AgentId"])

# Filling missing categorical values with "Unknown"
categorical_cols = ["OrganizationName", "MailingAddress1", "MailingCity", "PhysicalAddress1", "PhysicalCity"]
df_presenter_cleaned[categorical_cols] = df_presenter_cleaned[categorical_cols].fillna("Unknown")
artist_df_cleaned[["PhysicalAddressCity"]] = artist_df_cleaned[["PhysicalAddressCity"]].fillna("Unknown")

# Filling missing numerical values with median
numerical_cols = ["PhysicalGeoLatitude", "PhysicalGeoLongitude"]
df_presenter_cleaned[numerical_cols] = df_presenter_cleaned[numerical_cols].fillna(df_presenter_cleaned[numerical_cols].median())
artist_df_cleaned[numerical_cols] = artist_df_cleaned[numerical_cols].fillna(artist_df_cleaned[numerical_cols].median())

# Filling missing country/state IDs with a placeholder
df_presenter_cleaned[["MailingStateId", "MailingZip", "PhysicalStateId", "PhysicalZip"]] = \
    df_presenter_cleaned[["MailingStateId", "MailingZip", "PhysicalStateId", "PhysicalZip"]].fillna(-1)
artist_df_cleaned[["MailingCountryId", "PhysicalAddressStateId"]] = artist_df_cleaned[["MailingCountryId", "PhysicalAddressStateId"]].fillna(-1)

print("Missing values after preprocessing:")
print(df_presenter_cleaned.isnull().sum())
print(artist_df_cleaned.isnull().sum())
print(df_feedback_cleaned.isnull().sum())

Missing values after preprocessing:
PresenterId                 0
AgentId                     0
AccountName                 0
OrganizationName            0
PresenterTypeId             0
MailingAddress1             0
MailingAddress2         16547
MailingCity                 0
MailingStateId              0
MailingZip                  0
MailingCountryId            0
PhysicalAddress1            0
PhysicalAddress2        16547
PhysicalCity                0
PhysicalStateId             0
PhysicalZip                 0
PhysicalCountryId           0
PhysicalGeoLatitude         0
PhysicalGeoLongitude        0
IsActive                    0
CreatedDate                 0
CreatedById                 0
ParentPresenterId           0
dtype: int64
ArtistId                          0
Name                              0
MailingCountryId                  0
PhysicalAddressCity               0
PhysicalAddressStateId            0
PhysicalGeoLatitude               0
PhysicalGeoLongitude              0
IsActive 

In [14]:
# Convert feedback into numerical ratings
# Assuming IsNegative = False means a positive interaction (rating of 1), otherwise -1
df_feedback_cleaned["rating"] = np.where(df_feedback_cleaned["IsNegative"] == False, 1, -1)

# Create an interaction matrix
interaction_matrix = df_feedback_cleaned.pivot_table(
    index="PresenterPortalId", columns="ArtistId", values="rating", fill_value=0
)

# Checking sparsity again
sparsity = 1.0 - (interaction_matrix.count().sum() / (interaction_matrix.shape[0] * interaction_matrix.shape[1]))
print(f"Sparsity of Presenter-Artist Matrix: {sparsity * 100:.2f}%")

# Filter out presenters and artists with very few interactions (to reduce sparsity)
min_interactions = 3  # Set a minimum threshold
interaction_matrix = interaction_matrix.loc[(interaction_matrix != 0).sum(axis=1) >= min_interactions, :]
interaction_matrix = interaction_matrix.loc[:, (interaction_matrix != 0).sum(axis=0) >= min_interactions]

print(f"Shape of filtered interaction matrix: {interaction_matrix.shape}")

Sparsity of Presenter-Artist Matrix: 0.00%
Shape of filtered interaction matrix: (2, 0)


In [15]:
min_interactions = 1  # Allowing at least one interaction

interaction_matrix = df_feedback_cleaned.pivot_table(
    index="PresenterPortalId", columns="ArtistId", values="rating", fill_value=0
)

interaction_matrix = interaction_matrix.loc[(interaction_matrix != 0).sum(axis=1) >= min_interactions, :]
interaction_matrix = interaction_matrix.loc[:, (interaction_matrix != 0).sum(axis=0) >= min_interactions]

print(f"Updated shape of interaction matrix: {interaction_matrix.shape}")

Updated shape of interaction matrix: (47214, 4266)


In [16]:
from scipy.sparse import csr_matrix

# Reduce memory usage by converting types
df_bluecardartist["ArtistId"] = df_bluecardartist["ArtistId"].astype(np.int32)
df_bluecardartist["BlueCardId"] = df_bluecardartist["BlueCardId"].astype(np.int32)

# Assign a positive implicit rating for bookings
df_bluecardartist["implicit_rating"] = 1

# Create a sparse matrix for bookings
implicit_interactions = csr_matrix((
    df_bluecardartist["implicit_rating"].values, 
    (df_bluecardartist["BlueCardId"].values, df_bluecardartist["ArtistId"].values)
))

# Convert interaction_matrix to a sparse matrix
explicit_interactions = csr_matrix(interaction_matrix.values)

# Ensure both matrices have the same shape by padding
max_rows = max(explicit_interactions.shape[0], implicit_interactions.shape[0])
max_cols = max(explicit_interactions.shape[1], implicit_interactions.shape[1])

explicit_interactions.resize((max_rows, max_cols))
implicit_interactions.resize((max_rows, max_cols))

# Add explicit and implicit interactions
final_sparse_matrix = explicit_interactions + implicit_interactions

print(f"Final sparse interaction matrix shape: {final_sparse_matrix.shape}")

Final sparse interaction matrix shape: (869540, 104195)


In [20]:
from lightfm import LightFM
from lightfm.evaluation import precision_at_k

# Train-test split
from lightfm.data import Dataset

# Initialize dataset object
dataset = Dataset()

# Fit dataset with user-item interactions
dataset.fit(
    users=np.arange(final_sparse_matrix.shape[0]), 
    items=np.arange(final_sparse_matrix.shape[1])
)

# Convert sparse interaction matrix into LightFM format
(interactions, weights) = dataset.build_interactions(
    [(i, j) for i, j in zip(*final_sparse_matrix.nonzero())]
)

# Create model (Warp = better for implicit data, SVD-like factorization)
model = LightFM(loss="warp")  # Alternately, use loss="bpr" for SVD-like behavior

# Train the model
model.fit(interactions, epochs=10, num_threads=4)

# Evaluate the model
precision = precision_at_k(model, interactions, k=5).mean()
print(f"Precision at 5: {precision:.4f}")

Precision at 5: 0.0959


In [21]:
import pickle

# Save the trained LightFM model
with open("lightfm_model.pkl", "wb") as f:
    pickle.dump(model, f)

print("Model saved successfully!")

Model saved successfully!


In [25]:
import pickle

# Load model
with open("lightfm_model.pkl", "rb") as f:
    model = pickle.load(f)

print("Model loaded successfully!")

Model loaded successfully!


### Evaluations Before Tuning

In [26]:
from lightfm import LightFM
from lightfm.evaluation import precision_at_k

# Train-test split
from lightfm.data import Dataset

# Initialize dataset object
dataset = Dataset()

# Fit dataset with user-item interactions
dataset.fit(
    users=np.arange(final_sparse_matrix.shape[0]), 
    items=np.arange(final_sparse_matrix.shape[1])
)

# Convert sparse interaction matrix into LightFM format
(interactions, weights) = dataset.build_interactions(
    [(i, j) for i, j in zip(*final_sparse_matrix.nonzero())]
)

In [19]:
from lightfm.evaluation import precision_at_k, recall_at_k, auc_score
from sklearn.metrics import mean_squared_error

# Evaluate the model before tuning
print("\n--- Evaluating Model Before Hyperparameter Tuning ---")

# Compute Precision@5
precision_before = precision_at_k(model, interactions, k=5).mean()
print(f"Precision@5 (Before Tuning): {precision_before:.4f}")

# Compute Recall@5
recall_before = recall_at_k(model, interactions, k=5).mean()
print(f"Recall@5 (Before Tuning): {recall_before:.4f}")

# Compute AUC Score
auc_before = auc_score(model, interactions).mean()
print(f"AUC Score (Before Tuning): {auc_before:.4f}")


--- Evaluating Model Before Hyperparameter Tuning ---
Precision@5 (Before Tuning): 0.0959
Recall@5 (Before Tuning): 0.0819
AUC Score (Before Tuning): 0.9983


In [20]:
from sklearn.metrics import mean_squared_error
import numpy as np

# Extract non-zero actual interactions (only where users interacted with items)
actual_rows, actual_cols = final_sparse_matrix.nonzero()
actual_values = final_sparse_matrix.data  # Get actual ratings

# Get predictions for the same (presenter, artist) pairs
predicted_values = model.predict(actual_rows, actual_cols)

# Compute RMSE on only the non-zero interactions
rmse_before = np.sqrt(mean_squared_error(actual_values, predicted_values))
print(f"RMSE (Before Tuning): {rmse_before:.4f}")

RMSE (Before Tuning): 1.3549


In [21]:
# Get the top 5 presenters with most interactions
top_presenters = np.array(final_sparse_matrix.sum(axis=1)).flatten().argsort()[-5:][::-1]
 
print("Top 5 Presenters:", top_presenters)


Top 5 Presenters: [199444 840368 816921 801776 805039]


In [27]:
def get_recommendations(model, presenter_id, num_recommendations=5):

    num_artists = final_sparse_matrix.shape[1]  # Total number of artists

    # Ensure presenter_id is in an array
    presenter_id_array = np.array([presenter_id] * num_artists)  # Repeat presenter ID for all artists
    artist_id_array = np.arange(num_artists)  # Artist indices

    # Generate scores for all artists
    scores = model.predict(presenter_id_array, artist_id_array)

    # Get top N recommended artists
    top_artists = np.argsort(-scores)[:num_recommendations]

    return top_artists

# List of Top 5 Presenters
top_presenters = [199444, 840368, 816921, 801776, 805039]

# Get recommendations for top 5 presenters
recommendations = {presenter: get_recommendations(model, presenter) for presenter in top_presenters}

# Display results
for presenter, artists in recommendations.items():
    print(f"Recommended Artists for Presenter {presenter}: {artists}")

Recommended Artists for Presenter 199444: [  622  7426   687   133 16588]
Recommended Artists for Presenter 840368: [31050 32190 30663   296 31176]
Recommended Artists for Presenter 816921: [18325   315  2293  6542 19977]
Recommended Artists for Presenter 801776: [14265 17586 16632 11440   794]
Recommended Artists for Presenter 805039: [32190 31050   296 30663 31966]


In [23]:
# Create a mapping of ArtistId to Name
artist_id_to_name = dict(zip(artist_df["ArtistId"], artist_df["Name"]))

# Function to get artist names
def get_artist_names(artist_ids):
    return [artist_id_to_name.get(artist_id, "Unknown Artist") for artist_id in artist_ids]

# Get recommendations with names
recommendations_with_names = {
    presenter: get_artist_names(artists) for presenter, artists in recommendations.items()
}

# Display results with names
for presenter, artist_names in recommendations_with_names.items():
    print(f"Recommended Artists for Presenter {presenter}: {artist_names}")

Recommended Artists for Presenter 199444: ['TFC', 'TRADEMARK', 'THE GENTLEMEN & THEIR LADY', 'LIQUID PLEASURE WITH KENNY MANN', 'KLAXTON BROWN']
Recommended Artists for Presenter 840368: ['MICHEL JONS BAND', 'WE GOT THE BEAT', 'LOOSE CHAIN', 'THE VOLTAGE BROTHERS', 'SHIMMER BAND']
Recommended Artists for Presenter 816921: ['CIRQUE ZUMA ZUMA', 'BACKSTABBERS', "TERRY LEE & THE GT'S", 'THE SECOND CITY TOURING COMPANY', 'FAREWELL ANGELINA']
Recommended Artists for Presenter 801776: ['FANTASY', 'JAVA BAND', 'HOT SAUCE', "THE SWINGIN' RICHARDS", 'MR. POTATO HEAD']
Recommended Artists for Presenter 805039: ['WE GOT THE BEAT', 'MICHEL JONS BAND', 'THE VOLTAGE BROTHERS', 'LOOSE CHAIN', 'ASCENSION']


## Tuning

In [17]:
from lightfm.data import Dataset
import numpy as np
from scipy.sparse import csr_matrix

# Recreate LightFM Dataset
dataset = Dataset()

# Fit dataset with user-item interactions
dataset.fit(
    users=np.arange(final_sparse_matrix.shape[0]),  # All presenter IDs
    items=np.arange(final_sparse_matrix.shape[1])   # All artist IDs
)

# Convert sparse interaction matrix into LightFM format
(interactions, _) = dataset.build_interactions(
    [(i, j) for i, j in zip(*final_sparse_matrix.nonzero())]
)

print(f"Shape of interactions matrix: {interactions.shape}")

Shape of interactions matrix: (869540, 104195)


In [26]:
from lightfm import LightFM
from lightfm.evaluation import precision_at_k
import random

# Define search space
loss_functions = ["warp", "bpr"]
num_components_options = [10, 20, 50]
learning_rates = [0.01, 0.05, 0.1]
epochs = 5  # Initially low to test configurations faster
num_threads = 4  # Use 4 threads for parallel training

# Randomized Search (selects random hyperparameters)
num_trials = 5  
random_trials = [
    (random.choice(loss_functions), random.choice(num_components_options), random.choice(learning_rates))
    for _ in range(num_trials)
]

best_model = None
best_precision = 0
best_params = {}

for loss, num_components, lr in random_trials:
    print(f"Training model: loss={loss}, factors={num_components}, lr={lr}")

    # Train model
    model = LightFM(loss=loss, no_components=num_components, learning_rate=lr)
    model.fit(interactions, epochs=epochs, num_threads=num_threads)  # Set threads to 4

    # Evaluate model
    precision = precision_at_k(model, interactions, k=5).mean()
    print(f"Precision@5: {precision:.4f}")

    # Save best model
    if precision > best_precision:
        best_model = model
        best_precision = precision
        best_params = {"loss": loss, "factors": num_components, "lr": lr}

print("\nBest Model Found:")
print(f"Loss: {best_params['loss']}, Factors: {best_params['factors']}, LR: {best_params['lr']}")
print(f"Best Precision@5: {best_precision:.4f}")

Training model: loss=bpr, factors=10, lr=0.05
Precision@5: 0.2140
Training model: loss=bpr, factors=20, lr=0.05
Precision@5: 0.2202
Training model: loss=bpr, factors=50, lr=0.01
Precision@5: 0.0611
Training model: loss=warp, factors=10, lr=0.05
Precision@5: 0.0777
Training model: loss=warp, factors=20, lr=0.1
Precision@5: 0.1027

Best Model Found:
Loss: bpr, Factors: 20, LR: 0.05
Best Precision@5: 0.2202


In [25]:
from lightfm import LightFM
from lightfm.evaluation import precision_at_k

# Retrain the best model
best_model = LightFM(loss="bpr", no_components=20, learning_rate=0.05)
best_model.fit(interactions, epochs=30, num_threads=4)  # Train with more epochs

print("Model retrained successfully!")

Model retrained successfully!


In [34]:
import pickle

# Define the filename
model_filename = "lightfm_best_model.pkl"

# Save the model
with open(model_filename, "wb") as f:
    pickle.dump(best_model, f)

print(f"Model saved successfully as {model_filename}")

Model saved successfully as lightfm_best_model.pkl


In [20]:
import pickle

# Define the filename
model_filename = "lightfm_best_model.pkl"

# Load the saved model
with open(model_filename, "rb") as f:
    best_model = pickle.load(f)

print("Model loaded successfully!")

Model loaded successfully!


In [26]:
print("\n--- Evaluating Model After Hyperparameter Tuning ---")

# Compute Precision@5
precision_after = precision_at_k(best_model, interactions, k=5).mean()
print(f"Precision@5 (After Tuning): {precision_after:.4f}")

# Compute Recall@5
recall_after = recall_at_k(best_model, interactions, k=5).mean()
print(f"Recall@5 (After Tuning): {recall_after:.4f}")

# Compute AUC Score
auc_after = auc_score(best_model, interactions).mean()
print(f"AUC Score (After Tuning): {auc_after:.4f}")


--- Evaluating Model After Hyperparameter Tuning ---
Precision@5 (After Tuning): 0.3469
Recall@5 (After Tuning): 0.4945
AUC Score (After Tuning): 0.9962


In [22]:
import numpy as np
from sklearn.metrics import mean_squared_error

# Sample a subset of presenters to reduce memory load
num_samples = min(5000, final_sparse_matrix.shape[0])  # Pick 5000 or total if smaller
sampled_presenters = np.random.choice(final_sparse_matrix.shape[0], num_samples, replace=False)

# Get artist IDs
artist_ids = np.arange(final_sparse_matrix.shape[1])

# Store actual and predicted ratings
actual_ratings = []
predicted_ratings = []

# Process in batches
batch_size = 100  # Adjust based on available memory

for i in range(0, len(sampled_presenters), batch_size):
    batch_presenters = sampled_presenters[i : i + batch_size]  # Select batch

    # Convert to numpy array for consistent shape
    batch_actual_ratings = [final_sparse_matrix[presenter].toarray().flatten() for presenter in batch_presenters]
    batch_actual_ratings = np.array(batch_actual_ratings)

    # **Predict scores using batched presenter IDs**
    presenter_array = np.repeat(batch_presenters[:, np.newaxis], artist_ids.shape[0], axis=1)  # Expand presenters
    batch_predicted_ratings = np.array([best_model.predict(presenter_array[j], artist_ids) for j in range(len(batch_presenters))])

    # Store results
    actual_ratings.extend(batch_actual_ratings)
    predicted_ratings.extend(batch_predicted_ratings)

# Convert lists to numpy arrays
actual_ratings = np.array(actual_ratings)
predicted_ratings = np.array(predicted_ratings)

# Compute RMSE
rmse_after = np.sqrt(mean_squared_error(actual_ratings.flatten(), predicted_ratings.flatten()))
print(f"RMSE (After Tuning): {rmse_after:.4f}")

RMSE (After Tuning): 0.2542


In [31]:
# Create a mapping of ArtistId to Name
artist_id_to_name = dict(zip(artist_df["ArtistId"], artist_df["Name"]))

# Function to get artist names
def get_artist_names(artist_ids):
    return [artist_id_to_name.get(artist_id, "Unknown Artist") for artist_id in artist_ids]

# Get recommendations with names
recommendations_with_names = {
    presenter: get_artist_names(artists) for presenter, artists in recommendations.items()
}

# Display results with names
for presenter, artist_names in recommendations_with_names.items():
    print(f"Recommended Artists for Presenter {presenter}: {artist_names}")

def get_recommendations(model, presenter_id, num_recommendations=5):

    num_artists = final_sparse_matrix.shape[1]  # Total number of artists

    # Ensure presenter_id is in an array
    presenter_id_array = np.array([presenter_id] * num_artists)  # Repeat presenter ID for all artists
    artist_id_array = np.arange(num_artists)  # Artist indices

    # Generate scores for all artists
    scores = model.predict(presenter_id_array, artist_id_array)

    # Get top N recommended artists
    top_artists = np.argsort(-scores)[:num_recommendations]

    return top_artists

top_presenters = [199444, 840368, 816921, 801776, 805039]

# Get recommendations for top 5 presenters using the final trained model
final_recommendations = {presenter: get_recommendations(best_model, presenter) for presenter in top_presenters}

Recommended Artists for Presenter 199444: ['TFC', 'TRADEMARK', 'THE GENTLEMEN & THEIR LADY', 'LIQUID PLEASURE WITH KENNY MANN', 'KLAXTON BROWN']
Recommended Artists for Presenter 840368: ['MICHEL JONS BAND', 'WE GOT THE BEAT', 'LOOSE CHAIN', 'THE VOLTAGE BROTHERS', 'SHIMMER BAND']
Recommended Artists for Presenter 816921: ['CIRQUE ZUMA ZUMA', 'BACKSTABBERS', "TERRY LEE & THE GT'S", 'THE SECOND CITY TOURING COMPANY', 'FAREWELL ANGELINA']
Recommended Artists for Presenter 801776: ['FANTASY', 'JAVA BAND', 'HOT SAUCE', "THE SWINGIN' RICHARDS", 'MR. POTATO HEAD']
Recommended Artists for Presenter 805039: ['WE GOT THE BEAT', 'MICHEL JONS BAND', 'THE VOLTAGE BROTHERS', 'LOOSE CHAIN', 'ASCENSION']


In [32]:
import pickle
with open("lightfm_best_model.pkl", "wb") as f:
    pickle.dump(best_model, f)
print("Model saved successfully!")

Model saved successfully!
